File: bindings/dart/lib/src/engine.dart
import 'dart:convert';
import 'dart:ffi';
import 'dart:io';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import 'native.dart';

/// The base for every sone failure.
class SoneException implements Exception {
  SoneException(this.message);

  final String message;

  @override
  String toString() => '$runtimeType: $message';
}

/// The IR document could not be parsed.
class IrException extends SoneException {
  IrException(super.message);
}

/// A font or an image could not be loaded.
class AssetException extends SoneException {
  AssetException(super.message);
}

/// Layout or rasterization failed.
class RenderException extends SoneException {
  RenderException(super.message);
}

/// Owns the font registry and the decoded-image cache.
///
/// Skia's font collection is shared inside an engine, so one engine renders one
/// document at a time. FFI calls block the isolate, so a Flutter app should
/// render inside `Isolate.run` — with its own engine, not this one.
class Engine {
  /// [baseDir] is the directory relative asset paths resolve against.
  Engine([String? baseDir]) : _native = Native.instance {
    final dir = (baseDir ?? Directory.current.path).toNativeUtf8();
    try {
      _handle = _native.engineNew(dir);
    } finally {
      calloc.free(dir);
    }
    if (_handle == nullptr) {
      throw SoneException('could not create a sone engine');
    }
  }

  final Native _native;
  late final Pointer<Void> _handle;
  bool _closed = false;

  static Engine? _default;

  /// The process-wide engine, used when no explicit one is passed.
  static Engine get instance => _default ??= Engine();

  /// The native library version.
  static String get version => Native.instance.version().toDartString();

  bool get isClosed => _closed;

  void close() {
    if (_closed) return;
    _closed = true;
    _native.engineFree(_handle);
  }

  // ── fonts and assets ──────────────────────────────────────────────────────

  /// Register a font family from raw TTF/OTF bytes.
  void registerFont(String name, List<int> data) =>
      _withBytes(name, data, _native.registerFont);

  /// Register a font family from a file.
  void registerFontFile(String name, String path) {
    final namePointer = name.toNativeUtf8();
    final pathPointer = path.toNativeUtf8();
    try {
      _check(_native.registerFontFile(_live, namePointer, pathPointer));
    } finally {
      calloc.free(namePointer);
      calloc.free(pathPointer);
    }
  }

  /// Make bytes available to documents as `asset:name`.
  void registerImage(String name, List<int> data) =>
      _withBytes(name, data, _native.registerImage);

  /// Whether a family has been registered.
  bool hasFont(String name) {
    final pointer = name.toNativeUtf8();
    try {
      return _native.hasFont(_live, pointer);
    } finally {
      calloc.free(pointer);
    }
  }

  /// Every registered family name.
  List<String> fontFamilies() {
    final json =
        utf8.decode(_buffer((out) => _native.fontFamilies(_live, out)));
    return (jsonDecode(json) as List<dynamic>).cast<String>();
  }

  /// Drop every registered font.
  void resetFonts() => _native.resetFonts(_live);

  // ── rendering ─────────────────────────────────────────────────────────────

  /// Render an IR document to bytes.
  Uint8List render(
    String document, {
    OutputFormat format = OutputFormat.png,
    double? density,
    double quality = 1.0,
    bool strict = false,
  }) =>
      _call(document, format, density, quality, strict, (json, options) {
        return _buffer(
            (out) => _native.renderJson(_live, json, options, out));
      });

  /// One raster image per page. Requires `pageHeight` in the document config.
  List<Uint8List> renderPages(
    String document, {
    OutputFormat format = OutputFormat.png,
    double? density,
    double quality = 1.0,
    bool strict = false,
  }) =>
      _call(document, format, density, quality, strict, (json, options) {
        final list = calloc<SoneBufferList>();
        try {
          _check(_native.renderPages(_live, json, options, list));
          return List<Uint8List>.generate(list.ref.len, (index) {
            final page = list.ref.items[index];
            return Uint8List.fromList(page.data.asTypedList(page.len));
          });
        } finally {
          _native.bufferListFree(list);
          calloc.free(list);
        }
      });

  /// The computed layout tree, as JSON.
  String dumpLayout(String document) => _withDocument(
      document,
      (json) =>
          utf8.decode(_buffer((out) => _native.dumpLayout(_live, json, out))));

  /// Dataset-style metadata, as JSON.
  String dumpMetadata(String document, String granularity) =>
      _withDocument(document, (json) {
        final pointer = granularity.toNativeUtf8();
        try {
          return utf8.decode(_buffer(
              (out) => _native.dumpMetadata(_live, json, pointer, out)));
        } finally {
          calloc.free(pointer);
        }
      });

  // ── internals ─────────────────────────────────────────────────────────────

  Pointer<Void> get _live {
    if (_closed) throw StateError('this engine has been closed');
    return _handle;
  }

  /// Owns the document string and the options struct for exactly one call.
  T _call<T>(
      String document,
      OutputFormat format,
      double? density,
      double quality,
      bool strict,
      T Function(Pointer<Utf8>, Pointer<SoneRenderOptions>) body) {
    final options = calloc<SoneRenderOptions>();
    options.ref
      ..format = format.code
      // Zero tells the engine to fall back to the document's own config.
      ..density = density ?? 0.0
      ..quality = quality
      ..strict = strict ? 1 : 0;
    try {
      return _withDocument(document, (json) => body(json, options));
    } finally {
      calloc.free(options);
    }
  }

  T _withDocument<T>(String document, T Function(Pointer<Utf8>) body) {
    final json = document.toNativeUtf8();
    try {
      return body(json);
    } finally {
      calloc.free(json);
    }
  }

  void _withBytes(String name, List<int> data,
      int Function(Pointer<Void>, Pointer<Utf8>, Pointer<Uint8>, int) call) {
    final namePointer = name.toNativeUtf8();
    final buffer = calloc<Uint8>(data.isEmpty ? 1 : data.length);
    try {
      buffer.asTypedList(data.length).setAll(0, data);
      _check(call(_live, namePointer, buffer, data.length));
    } finally {
      calloc.free(namePointer);
      calloc.free(buffer);
    }
  }

  Uint8List _buffer(int Function(Pointer<SoneBuffer>) call) {
    final out = calloc<SoneBuffer>();
    try {
      _check(call(out));
      final buffer = out.ref;
      if (buffer.data == nullptr || buffer.len == 0) return Uint8List(0);
      return Uint8List.fromList(buffer.data.asTypedList(buffer.len));
    } finally {
      _native.bufferFree(out);
      calloc.free(out);
    }
  }

  void _check(int status) {
    if (status == SoneStatus.ok) return;
    final pointer = _native.lastError(_handle);
    final message = pointer == nullptr
        ? 'sone failed with status $status'
        : pointer.toDartString();
    throw switch (status) {
      SoneStatus.invalidArgument => ArgumentError(message),
      SoneStatus.irError => IrException(message),
      SoneStatus.assetError => AssetException(message),
      _ => RenderException(message),
    };
  }
}

/// Font registration on the process-wide engine, for scripts that do not want
/// to own one. Skia carries no system fonts, so at least one family must be
/// registered before any text renders.
abstract final class Font {
  static void load(String name, String path) =>
      Engine.instance.registerFontFile(name, path);

  static void loadBytes(String name, List<int> data) =>
      Engine.instance.registerFont(name, data);

  static bool has(String name) => Engine.instance.hasFont(name);

  static List<String> families() => Engine.instance.fontFamilies();

  static void reset() => Engine.instance.resetFonts();
}

File: bindings/php/src/Engine.php
<?php

declare(strict_types=1);

namespace Sone;

/**
 * Owns the font registry and the decoded-image cache.
 *
 * Skia's font collection is shared inside an engine, so one engine renders one
 * document at a time. PHP has no threads to guard against, but the rule still
 * holds for anything that forks a request into workers.
 */
final class Engine
{
    private \FFI $ffi;
    private \FFI\CData $handle;
    private bool $closed = false;

    private static ?self $default = null;

    /** @param string|null $baseDir the directory relative asset paths resolve against */
    public function __construct(?string $baseDir = null)
    {
        $this->ffi = Native::ffi();
        $handle = $this->ffi->sone_engine_new($baseDir ?? getcwd());
        if ($handle === null) {
            throw new SoneException('could not create a sone engine');
        }
        $this->handle = $handle;
    }

    public function __destruct()
    {
        $this->close();
    }

    /** The process-wide engine, used when no explicit one is passed. */
    public static function default(): self
    {
        return self::$default ??= new self();
    }

    public static function version(): string
    {
        return Native::ffi()->sone_version();
    }

    public function close(): void
    {
        if (!$this->closed) {
            $this->closed = true;
            $this->ffi->sone_engine_free($this->handle);
        }
    }

    // ── fonts and assets ────────────────────────────────────────────────────

    /** Register a font family from raw TTF/OTF bytes. */
    public function registerFont(string $name, string $data): void
    {
        $this->withBytes($data, fn (\FFI\CData $buffer, int $length) =>
            $this->check($this->ffi->sone_register_font($this->live(), $name, $buffer, $length)));
    }

    /** Register a font family from a file. */
    public function registerFontFile(string $name, string $path): void
    {
        $this->check($this->ffi->sone_register_font_file($this->live(), $name, $path));
    }

    /** Make bytes available to documents as `asset:name`. */
    public function registerImage(string $name, string $data): void
    {
        $this->withBytes($data, fn (\FFI\CData $buffer, int $length) =>
            $this->check($this->ffi->sone_register_image($this->live(), $name, $buffer, $length)));
    }

    public function hasFont(string $name): bool
    {
        return $this->ffi->sone_has_font($this->live(), $name);
    }

    /** @return list<string> */
    public function fontFamilies(): array
    {
        return json_decode($this->buffer(fn (\FFI\CData $out) =>
            $this->ffi->sone_font_families($this->live(), $out)), true);
    }

    public function resetFonts(): void
    {
        $this->ffi->sone_reset_fonts($this->live());
    }

    // ── rendering ───────────────────────────────────────────────────────────

    /** Render an IR document to bytes. */
    public function render(
        string $document,
        OutputFormat|string $format = OutputFormat::Png,
        ?float $density = null,
        float $quality = 1.0,
        bool $strict = false,
    ): string {
        $options = $this->options($format, $density, $quality, $strict);

        return $this->buffer(fn (\FFI\CData $out) =>
            $this->ffi->sone_render_json($this->live(), $document, \FFI::addr($options), $out));
    }

    /**
     * One raster image per page. Requires `pageHeight` in the document config.
     *
     * @return list<string>
     */
    public function renderPages(
        string $document,
        OutputFormat|string $format = OutputFormat::Png,
        ?float $density = null,
        float $quality = 1.0,
        bool $strict = false,
    ): array {
        $options = $this->options($format, $density, $quality, $strict);
        $list = $this->ffi->new('SoneBufferList');
        try {
            $this->check($this->ffi->sone_render_pages(
                $this->live(),
                $document,
                \FFI::addr($options),
                \FFI::addr($list),
            ));
            $pages = [];
            for ($index = 0; $index < $list->len; $index++) {
                $page = $list->items[$index];
                $pages[] = $page->len > 0 ? \FFI::string($page->data, $page->len) : '';
            }

            return $pages;
        } finally {
            $this->ffi->sone_buffer_list_free(\FFI::addr($list));
        }
    }

    /** The computed layout tree, as JSON. */
    public function dumpLayout(string $document): string
    {
        return $this->buffer(fn (\FFI\CData $out) =>
            $this->ffi->sone_dump_layout($this->live(), $document, $out));
    }

    /** Dataset-style metadata, as JSON. */
    public function dumpMetadata(string $document, Granularity|string $granularity = Granularity::Node): string
    {
        $name = $granularity instanceof Granularity ? $granularity->value : $granularity;

        return $this->buffer(fn (\FFI\CData $out) =>
            $this->ffi->sone_dump_metadata($this->live(), $document, $name, $out));
    }

    // ── internals ───────────────────────────────────────────────────────────

    private function live(): \FFI\CData
    {
        if ($this->closed) {
            throw new SoneException('this engine has been closed');
        }

        return $this->handle;
    }

    /**
     * The options struct, always handed over as a pointer.
     *
     * It used to be passed by value, which segfaulted PHP on Linux x86-64 while
     * working on macOS and Windows — the C ABI takes a pointer now.
     */
    private function options(OutputFormat|string $format, ?float $density, float $quality, bool $strict): \FFI\CData
    {
        $name = strtolower($format instanceof OutputFormat ? $format->value : $format);
        $code = Native::FORMATS[$name] ?? throw new \InvalidArgumentException("unknown output format \"$name\"");

        $options = $this->ffi->new('SoneRenderOptions');
        $options->format = $code;
        // Zero tells the engine to fall back to the document's own config.
        $options->density = $density ?? 0.0;
        $options->quality = $quality;
        $options->strict = $strict ? 1 : 0;

        return $options;
    }

    private function withBytes(string $data, \Closure $body): void
    {
        $length = \strlen($data);
        // Instance calls, not FFI::new()/FFI::cast(): the static forms are
        // deprecated as of PHP 8.5.
        $buffer = $this->ffi->new('uint8_t[' . max($length, 1) . ']', false);
        try {
            if ($length > 0) {
                \FFI::memcpy($buffer, $data, $length);
            }
            $body($this->ffi->cast('uint8_t*', $buffer), $length);
        } finally {
            \FFI::free($buffer);
        }
    }

    private function buffer(\Closure $call): string
    {
        $out = $this->ffi->new('SoneBuffer');
        try {
            $this->check($call(\FFI::addr($out)));

            return $out->len > 0 ? \FFI::string($out->data, $out->len) : '';
        } finally {
            $this->ffi->sone_buffer_free(\FFI::addr($out));
        }
    }

    private function check(int $status): void
    {
        if ($status === Native::OK) {
            return;
        }
        $pointer = $this->ffi->sone_engine_last_error($this->handle);
        $message = $pointer === null ? "sone failed with status $status" : $pointer;

        throw match ($status) {
            Native::INVALID_ARGUMENT => new \InvalidArgumentException($message),
            Native::IR_ERROR => new IrException($message),
            Native::ASSET_ERROR => new AssetException($message),
            default => new RenderException($message),
        };
    }
}

/**
 * Font registration on the process-wide engine.
 *
 * Skia carries no system fonts, so at least one family must be registered
 * before any text renders.
 */
final class Font
{
    public static function load(string $name, string $path): void
    {
        Engine::default()->registerFontFile($name, $path);
    }

    public static function loadBytes(string $name, string $data): void
    {
        Engine::default()->registerFont($name, $data);
    }

    public static function has(string $name): bool
    {
        return Engine::default()->hasFont($name);
    }

    /** @return list<string> */
    public static function families(): array
    {
        return Engine::default()->fontFamilies();
    }

    public static function reset(): void
    {
        Engine::default()->resetFonts();
    }
}
